service/history/workflow/mutable_state_impl.go

```diff
@@ -202,10 +202,12 @@ type (
 		transitionHistoryEnabled bool
 		// following fields are for tracking if sub state machines are updated
 		// in a transaction. They are similar to field like updateActivityInfos
-		visibilityUpdated     bool
-		executionStateUpdated bool
-		workflowTaskUpdated   bool
-		updateInfoUpdated     map[string]struct{}
+		visibilityUpdated                  bool
+		executionStateUpdated              bool
+		workflowTaskUpdated                bool
+		updateInfoUpdated                  map[string]struct{}
+		updateActivityInfosUserDataUpdated map[int64]struct{}
+		updateTimerInfosUserDataUpdated    map[string]struct{}
```

Comment:
nit: 
```suggestion
		activityInfosUserDataUpdated map[int64]struct{}
		timerInfosUserDataUpdated    map[string]struct{}
```

Reply:
Plz also add some comment explaining why we need this in addition to `updateActivityInfos` and `updateTimerInfos`